sql/schemachanger: set schema_locked automatically for declarative

fqazi opened the pull request:
Previously, the schema_locked setting had to be manually toggled by the user before executing DDL on a table. This meant that setting / unsetting schema_locked was a manual process for users. The declarative schema changer makes it much easier to tie setting / unsetting the locked setting within the schema change path. This patch adds a new TRANSIENT_PUBLIC state that allows elements to be toggled. The builder is modified to emit this schema_locked element in a TRANSIENT_PUBLIC state.

Informs: #129694
Release note: None

fqazi committed:
sql/schemachanger: set schema_locked automatically for declarative

Previously, the schema_locked setting had to be manually toggled by the
user before executing DDL on a table. This meant that setting /
unsetting schema_locked was a manual process for users. The declarative
schema changer makes it much easier to tie setting / unsetting the
locked setting within the schema change path. This patch adds a new
TRANSIENT_PUBLIC state that allows elements to be toggled. The builder
is modified to emit this schema_locked element in a TRANSIENT_PUBLIC
state.

Informs: #129694
Release note: None

rafiss reviewed (changes requested):
this is looking pretty good! mostly had minor comments.

the next step after this should be to modify CREATE TABLE to set schema_locked automatically, but that should be a separate PR.

> Reviewed 31 of 72 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143273#-:-OMEM5_n48h2_9iPoGl1:b5egx1h)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/sql/schemachanger/scplan/internal/rules/helpers.go` line 287 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143273#-OMELJgQ1FwrUkOiRxaB:-OMELJgQ1FwrUkOiRxaC:b-t8x5rv) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c3ffb7137740f113f7b24ddf925b37c29e3dbe4/pkg/sql/schemachanger/scplan/internal/rules/helpers.go#L287)):*
> ```Go
> 
> 	toTransientPublicToDropUntyped = screl.Schema.Def2(
> 		"toTransientPublicToDrop",
> ```

i'm not really following the "to X to Y" names for this clause and the below one. should it say "or" instead of "to"

or perhaps some comments on these variables would help me understand what they are used for.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table.go` line 127 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143273#-OMEHkVO9ltAGGPAIiz7:-OMEHkVP0EWzKhbUwhyN:b-6115lq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c3ffb7137740f113f7b24ddf925b37c29e3dbe4/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table.go#L127)):*
> ```Go
> 	}
> 	checkTableSchemaChangePrerequisites(b, elts, n)
> 	if schemaLockedElts := elts.FilterTableSchemaLocked().MustHaveZeroOrOne(); !schemaLockedElts.IsEmpty() {
> ```

it looks like this is handled as part of `checkTableSchemaChangePrerequisites`, so IIUC we shouldn't need the DropTransient here also.
___
*[`pkg/sql/schemachanger/scpb/constants.go` line 50 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143273#-OMEIof25rNR482QNjXi:-OMEIof25rNR482QNjXj:btuccvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c3ffb7137740f113f7b24ddf925b37c29e3dbe4/pkg/sql/schemachanger/scpb/constants.go#L50)):*
> ```Go
> 	// existence before disappearing again. Otherwise, an element whose current
> 	// and target statuses are both ABSENT won't experience any state transitions.
> 	Transient TargetStatus = TargetStatus(Status_TRANSIENT_ABSENT)
> ```

nit: in a separate commit, let's rename this to `TransientAbsent`
___
*[`pkg/sql/schemachanger/scpb/constants.go` line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143273#-OMEIE3Q-aBIWCFgJb9Y:-OMEIE3Q-aBIWCFgJb9Z:bugygq4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c3ffb7137740f113f7b24ddf925b37c29e3dbe4/pkg/sql/schemachanger/scpb/constants.go#L52)):*
> ```Go
> 	Transient TargetStatus = TargetStatus(Status_TRANSIENT_ABSENT)
> 
> 	TransientPublic TargetStatus = TargetStatus(Status_TRANSIENT_PUBLIC)
> ```

can you add a comment on this; it's a pretty special kind of status, so worth including some details
___
*[`pkg/sql/schemachanger/scplan/internal/opgen/register.go` line 101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143273#-OMEJZWe2ETtlv2KVCZU:-OMEJZWe2ETtlv2KVCZV:b-tugz0b) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c3ffb7137740f113f7b24ddf925b37c29e3dbe4/pkg/sql/schemachanger/scplan/internal/opgen/register.go#L101)):*
> ```Go
> 			p = &transientSpec
> 		case scpb.Status_TRANSIENT_PUBLIC:
> 			p = &transientAddSpec
> ```

am i thinking about this backwards? i feel like it should be:
```
		case scpb.Status_TRANSIENT_ABSENT:
			p = &transientAddSpec
		case scpb.Status_TRANSIENT_PUBLIC:
			p = &transientDropSpec
```

(either way, let's rename the `transientSpec` variable to include "Drop" in it)
___
*[`pkg/sql/schemachanger/scpb/state.go` line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143273#-OMEIPiZ64iS-UQL_X9W:-OMEIPiZ64iS-UQL_X9X:bs4nrzy) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c3ffb7137740f113f7b24ddf925b37c29e3dbe4/pkg/sql/schemachanger/scpb/state.go#L102)):*
> ```Go
> 		// Transient public elements always have the end goal
> 		// of always returning back to public. So, the target
> 		// always stays the same (i.e. turn schema locked again
> ```

it should say "turn schema_locked off and on again"


<!-- Sent from Reviewable.io -->

fqazi committed:
sql/schemachanger: rename Transient to TransientAbsent

Now that we have added TransientPublic, we should rename Transient to
TransientAbsent for clarity.

Release note: None

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143273)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table.go` line 127 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143273#-OMEHkVO9ltAGGPAIiz7:-OMHlRYUDP1uXahzOWjF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c3ffb7137740f113f7b24ddf925b37c29e3dbe4/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table.go#L127)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

it looks like this is handled as part of `checkTableSchemaChangePrerequisites`, so IIUC we shouldn't need the DropTransient here also.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scpb/constants.go` line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143273#-OMEIE3Q-aBIWCFgJb9Y:-OMHm7B-5-q5hkWrOhaI:bk95cb3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c3ffb7137740f113f7b24ddf925b37c29e3dbe4/pkg/sql/schemachanger/scpb/constants.go#L52)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

can you add a comment on this; it's a pretty special kind of status, so worth including some details
</blockquote></details>

Added the comment:

// TransientPublic is similar to ToPublic in that the element should persist after// the schema change is completed. The element is assumed to exist before the schema  
// change; therefore, this target status is used to ensure it becomes absent and  
// becomes public at a later stage.
___
*[`pkg/sql/schemachanger/scpb/state.go` line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143273#-OMEIPiZ64iS-UQL_X9W:-OMHn3k76iV-Mt8cJ24c:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c3ffb7137740f113f7b24ddf925b37c29e3dbe4/pkg/sql/schemachanger/scpb/state.go#L102)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

it should say "turn schema_locked off and on again"
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scplan/internal/opgen/register.go` line 101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143273#-OMEJZWe2ETtlv2KVCZU:-OMIh3rvD3RzGWB5tNw4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c3ffb7137740f113f7b24ddf925b37c29e3dbe4/pkg/sql/schemachanger/scplan/internal/opgen/register.go#L101)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

am i thinking about this backwards? i feel like it should be:
```
		case scpb.Status_TRANSIENT_ABSENT:
			p = &transientAddSpec
		case scpb.Status_TRANSIENT_PUBLIC:
			p = &transientDropSpec
```

(either way, let's rename the `transientSpec` variable to include "Drop" in it)
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/helpers.go` line 287 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143273#-OMELJgQ1FwrUkOiRxaB:-OMIgX_t55vXLvVY-Md6:bc4qhrs) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c3ffb7137740f113f7b24ddf925b37c29e3dbe4/pkg/sql/schemachanger/scplan/internal/rules/helpers.go#L287)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i'm not really following the "to X to Y" names for this clause and the below one. should it say "or" instead of "to"

or perhaps some comments on these variables would help me understand what they are used for.
</blockquote></details>

Added comments on both and renamed them describe what they are checking:

// toDropToTransientPublicUntyped makes sure target1 is targeting// DROP/TRANSIENT\_DROP and target2 is targeting TRANSIENT\_PUBLIC.

// toPublicToTransientPublicUntyped makes sure target1 is targeting// PUBLIC and target2 is targeting TRANSIENT\_PUBLIC.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
great work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143273#-:-OMn-Ovj73afNAeuXyD2:by276bc)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

fqazi commented:
@rafiss TFTR!

bors r+

fqazi commented:
bors retry

Needed a rebase to address a missing cluster version import
